plugins/steelseries/fu-steelseries-sonic.c

```diff
@@ -0,0 +1,1238 @@
+	if (endpoints == NULL || endpoints->len != 1) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "endpoint not found");
+		return FALSE;
+	}
+
+	ep = g_ptr_array_index(endpoints, 0);
+	ep_id = g_usb_endpoint_get_address(ep);
+	packet_size = g_usb_endpoint_get_maximum_packet_size(ep);
+
+	self->iface_idx = iface_id;
+	self->ep = ep_id;
+	self->in_size = packet_size;
+
+	fu_usb_device_add_interface(FU_USB_DEVICE(self), iface_id);
+
+	/* success */
+	return TRUE;
+#else
+	/* failed */
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "this version of GUsb is not supported");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_steelseries_sonic_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	SteelseriesSonicChip chip;
+	g_autoptr(FwupdRequest) request = NULL;
+	g_autofree gchar *msg = NULL;
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 50);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 50);
+
+	/* mouse */
+	chip = STEELSERIES_SONIC_CHIP_MOUSE;
+	if (!fu_steelseries_sonic_restart(device, chip, fu_progress_get_child(progress), error)) {
+		g_prefix_error(error, "failed to restart chip %u: ", chip);
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* dongle (nordic, holtek; same command) */
+	chip = STEELSERIES_SONIC_CHIP_HOLTEK;
+	if (!fu_steelseries_sonic_restart(device, chip, fu_progress_get_child(progress), error)) {
+		g_prefix_error(error, "failed to restart chip %u: ", chip);
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* the user has to do something */
+	msg = g_strdup_printf("%s needs to be manually restarted to complete the update. "
+			      "Please unplug the dongle and then re-plug it.",
+			      fu_device_get_name(device));
+	request = fwupd_request_new();
+	fwupd_request_set_kind(request, FWUPD_REQUEST_KIND_IMMEDIATE);
+	fwupd_request_set_id(request, FWUPD_REQUEST_ID_REMOVE_REPLUG);
+	fwupd_request_set_message(request, msg);
+	fu_device_emit_request(device, request);
+
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_sonic_prepare(FuDevice *device,
+			     FuProgress *progress,
+			     FwupdInstallFlags flags,
+			     GError **error)
+{
+	SteelseriesSonicWirelessStatus wl_status;
+	guint16 bat_state;
+
+	if (!fu_steelseries_sonic_wireless_status(device, &wl_status, error)) {
+		g_prefix_error(error, "failed to get wireless status: ");
+		return FALSE;
+	}
+	g_debug("WirelessStatus: %u", wl_status);
+	if (wl_status != STEELSERIES_SONIC_WIRELESS_STATE_CONNECTED) {
+		g_autoptr(FwupdRequest) request = NULL;
+		g_autoptr(GTimer) timer = NULL;
+		g_autofree gchar *msg = NULL;
+
+		/* the user has to do something */
+		msg = g_strdup_printf("%s needs to be connected to start the update. "
+				      "Please press any button.",
```

Comment:
i think this needs to be a bit more specific. A button on _what_?

Reply:
Ah ah, I tried to be the most generic in case of the the plugin is reuse by device such as keyboard or headset.

So okay. I will be more specific, et even more if more products are added in the future.